planner/core/planbuilder.go

```diff
@@ -3215,7 +3215,8 @@ func (b *PlanBuilder) buildSetValuesOfInsert(ctx context.Context, insert *ast.In
 	insertPlan.AllAssignmentsAreConstant = true
 	for i, assign := range insert.Setlist {
 		defaultExpr := extractDefaultExpr(assign.Expr)
-		if defaultExpr != nil {
+		if defaultExpr != nil && !defaultExpr.NameIsGiven {
+			// Set the name of the column in the assignment
```

Comment:
In `buildValuesListOfInsert`, there is a switch-case containing `DefaultExpr`, `ValueExpr`, `ParamMarkerExpr`, and `default`, but there is not such switch-case in `buildSetValuesOfInsert`. I don't know why. Did you figure out why?

Reply:
I guess `DefaultExpr.Name` was previously intended to distinguish `default(a)` from `default`. It's better to follow the convention and call `b.getDefaultValue(exprCols[i])` to return another expression like what `buildValuesListOfInsert` does.

Reply:
Thanks for the suggestion.
I reverted the change in the parser and reorganised the code a bit for `buildValuesListOfInsert`/`buildSetValuesOfInsert` but then I did some more testing and found issues with `UPDATE` and `INSERT ON DUPLICATE KEY UPDATE` as well, and decided to break the "bug" compatibility with MySQL, by also making `SELECT DEFAULT(timestamp_col_with_CURRENT_TIMESTAMP_default)` to return `NOW()` which is more logical and matches INSERT/UPDATE behaviour.

Reply:
> and decided to break the "bug" compatibility with MySQL, by also making SELECT DEFAULT(timestamp_col_with_CURRENT_TIMESTAMP_default) to return NOW() which is more logical and matches INSERT/UPDATE behaviour.

So you think this is a bug:

```sql
mysql> create table t(id int, ti timestamp default current_timestamp);
Query OK, 0 rows affected (0.37 sec)

mysql> insert t(id) value(1);
Query OK, 1 row affected (0.02 sec)

mysql> select default(ti) from t;
+-------------+
| default(ti) |
+-------------+
| NULL        |
+-------------+
1 row in set (0.01 sec)
```

I'm not sure about this. WDYT? @bb7133 @morgo

Reply:
Yes, I think that is a bug. My guess is that it was implemented this way, to be backwards compatible *within* MySQL, since before the DEFAULT CURRENT_TIMESTAMP, it would automatically make the first timestamp to convert 0 or NULL values to current @@timestamp/NOW(). I.e. that is why it would internally return NULL, to make the already existing code to convert it to @@timestamp/NOW().

My main reasoning for consider `SELECT DEFAULT(ts_col)` returning `NULL` is a bug, is that it is not the same as `INSERT INTO t VALUES () / (DEFAULT) / DEFAULT(ts_col)` which is what I would expect. Since INSERT/UPDATE would set the value to @@timestamp or NOW(), that is what I expect SELECT to return as well.

Reply:
I agree it looks like a bug - but when we are not sure, we can file bugs against MySQL and wait for them to be verified. It benefits both of us :-)

Reply:
Reported https://bugs.mysql.com/bug.php?id=105829

Reply:
MySQL deemed bug#105829 as 'Not a bug', claiming it is explained here:
https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html
Which I cannot see any mentioning of how it handles `DEFAULT(col)`.

I added another comment just in the MySQL bug to clarify my expectations.
```
I could find a part that may be appropriate in 11.6:
"However, the use of DEFAULT(col_name) to specify the default value for a named column is permitted only for columns that have a literal default value, not for columns that have an expression default value."

Which is clear with this example:
create table t (a int primary key, b int default (3 + 2));

insert into t (a) values (1);
mysql> select default(b) from t;
ERROR 3773 (HY000): DEFAULT function cannot be used with default value expressions

But then I would expect an error when doing it with 'b timestamp default current_timestamp' too, and not silently insert NULL for the second row in the 'How to repeat' section!
```

Since TiDB does not yet support Expressions for column DEFAULTS, see https://github.com/pingcap/tidb/issues/30553, we can safely ignore everything else than `CURRENT_TIMESTAMP` and `NOW()`. (I will also ignore generated columns for now :) )


So current master vs MySQL 8.0.27 does
```
create table t (a int primary key, b timestamp default current_timestamp);
insert into t values (1,default),(2,default(b));
insert into t set a = 3, b = default;
insert into t set a = 4, b = default(b);
```
TiDB:
```
tidb> select a,b,default(b) from t;
+---+---------------------+---------------------+
| a | b                   | default(b)          |
+---+---------------------+---------------------+
| 1 | 2021-12-08 18:00:25 | 0000-00-00 00:00:00 |
| 2 | 2021-12-08 18:00:25 | 0000-00-00 00:00:00 |
| 3 | 0000-00-00 00:00:00 | 0000-00-00 00:00:00 |
| 4 | 0000-00-00 00:00:00 | 0000-00-00 00:00:00 |
+---+---------------------+---------------------+
```
MySQL:
```
mysql> select a,b,default(b) from t;
+---+---------------------+------------+
| a | b                   | default(b) |
+---+---------------------+------------+
| 1 | 2021-12-08 18:01:44 | NULL       |
| 2 | NULL                | NULL       |
| 3 | 2021-12-08 18:01:55 | NULL       |
| 4 | NULL                | NULL       |
+---+---------------------+------------+
```
And with current state of the patch it would be:
```
tidb> select a,b,default(b) from t;
+---+---------------------+---------------------+
| a | b                   | default(b)          |
+---+---------------------+---------------------+
| 1 | 2021-12-08 18:05:41 | 2021-12-08 18:05:52 |
| 2 | 2021-12-08 18:05:41 | 2021-12-08 18:05:52 |
| 3 | 2021-12-08 18:05:41 | 2021-12-08 18:05:52 |
| 4 | 2021-12-08 18:05:43 | 2021-12-08 18:05:52 |
+---+---------------------+---------------------+
4 rows in set (0,00 sec)
```

@morgo @djshow832 should we be bug compatible with MySQL when it comes to handling `DEFAULT(col)` when col is a timestamp and DEFAULT is `CURRENT_TIMESTAMP`/`NOW()` ? If so, I think it makes sense to create a new bug for that and consider this bug to handle the `INSERT ... VALUES` vs `INSERT ... SET` difference.